### PR TITLE
Concurrent Editing Engine — Replacing Edit Leases with Hybrid OT

### PR DESCRIPTION
**Pull request by HamptonMakes**
## Summary

Implements the full [Concurrent Editing Engine plan](http://localhost:3000/plans/0f18089e-ad16-41d3-82f3-f25ffc47a574) — replacing exclusive edit leases with a hybrid OT system that supports concurrent, non-blocking editing by multiple agents.

### Phase 1: Position Resolution Engine
- **`Plans::PositionResolver`** — resolves semantic operations (`replace_exact`, `insert_under_heading`, `delete_paragraph_containing`) to precise character ranges
- `replace_exact` gains `occurrence` (target Nth match) and `replace_all` (replace every match) params, replacing the confusing `count` semantics
- All applied operations now include `resolved_range`, `new_range`, and `delta` in `operations_json` for future OT transforms

### Phase 2: Edit Sessions
- **`EditSession`** model — lightweight, non-exclusive sessions that allow multiple agents to edit simultaneously
- **`Plans::CommitSession`** — batches all operations in a session into a single `PlanVersion` (solves version churn)
- **`CommitExpiredSessionJob`** — auto-commits expired sessions via SolidQueue
- API: `POST /sessions`, `GET /sessions/:id`, `POST /sessions/:id/commit`

### Phase 3: Auto-Rebase / Transform
- **`Plans::TransformRange`** — OT engine that transforms character ranges through intervening edits
- Operations controller now supports **3 modes**: session-based, legacy lease, and direct (lease-free with auto-rebase)
- Stale `base_revision` triggers automatic rebase; conflicts only on true position overlap (not just stale revision number)
- Row-level locking (`plan.lock!`) prevents race conditions on concurrent version creation

### Phase 4: Comment Anchor Tracking
- Comments resolve `anchor_text` to character positions on creation
- `mark_out_of_date_for_new_version!` now uses position-based range transforms — comments only go outdated when their specific anchored text is modified
- Enhanced `GET /comments` response with `anchor_context` (highlighted surrounding text) and `anchor_valid`

### Phase 5: Transition Support
- Legacy lease API (`POST /lease`, `DELETE /lease`, operations with `lease_token`) remains fully functional
- Both APIs coexist; old API can be removed once all agents migrate

## Key Design Decisions

| Decision | Rationale |
|----------|-----------|
| Pre-resolved ranges passed through OT pipeline | Avoids re-resolving positions after rebase (caught in code review as critical bug) |
| Row-level DB locks on version creation | Prevents duplicate revision numbers under concurrency |
| Session TTL: 10min local / 30min cloud persona | Cloud personas are server-side with no network latency |
| Transform chain capped at 20 revisions | Keeps transform cost bounded |
| Partial commit fails entirely | A committed session always represents a complete, coherent set of changes |

## Testing

**398 tests, 0 failures** (155 new tests added)

- `spec/services/plans/position_resolver_spec.rb` — 38 tests covering all operation types, edge cases, occurrence/replace_all
- `spec/services/plans/transform_range_spec.rb` — 38 tests including commutativity, chaining, zero-width ranges
- `spec/services/plans/commit_session_spec.rb` — 16 tests including concurrent session scenarios
- `spec/services/plans/apply_operations_spec.rb` — 27 tests (8 new for position data and new params)
- `spec/models/edit_session_spec.rb` — 24 tests for model lifecycle
- `spec/models/comment_thread_anchor_spec.rb` — 11 tests for position-based outdating
- `spec/requests/api/v1/sessions_spec.rb` — 14 tests for all API modes
- `spec/jobs/commit_expired_session_job_spec.rb` — 6 tests for auto-commit behavior

## Database Changes

- **New table**: `edit_sessions` (UUID PK, FK to plans/organizations/plan_versions)
- **Modified**: `comment_threads` gains `anchor_start`, `anchor_end`, `anchor_revision` columns

### CODE REVIEW

**Comment by HamptonMakes** [comment]
@codex please re-review

File: app/controllers/api/v1/operations_controller.rb [hunk 1]
```diff
@@ -24,6 +18,64 @@ def create
...
+          apply_direct(operations, base_revision)
+        end
+      rescue Plans::OperationError => e
+        render json: { error: e.message }, status: :unprocessable_entity
+      end
+
+      private
+
+      def apply_with_session(operations, base_revision)
+        session = @plan.edit_sessions.find_by(id: params[:session_id])
```
[the remark this answers is not in the document]
Comment (HamptonMakes) [reply app/controllers/api/v1/operations_controller.rb]: Fixed in a2c0065. Session lookups in both `OperationsController` and `SessionsController` now include `actor_id: @api_token.id`, so a token can only access sessions it created. Prevents cross-actor session hijacking and attribution spoofing.

File: app/services/plans/commit_session.rb [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+    def call
+      raise SessionNotOpenError, "Session is not open" unless @session.open?
+
+      plan = @session.plan
+
+      # No operations: just mark committed, no version created
+      unless @session.has_operations?
+        @session.update!(status: "committed", committed_at: Time.current)
+        return { session: @session, version: nil }
```
[the remark this answers is not in the document]
Comment (HamptonMakes) [reply app/services/plans/commit_session.rb]: Fixed in a2c0065. The empty-session fast-path now runs inside the transaction after `@session.lock!` + `open?` recheck. This prevents a concurrent `POST /operations` from appending ops between the `has_operations?` check and the status update, which would have stranded newly added edits.

File: app/models/comment_thread.rb [hunk 3]
```diff
@@ -12,26 +14,52 @@ class CommentThread < ApplicationRecord
...
-      if thread.anchor_context.present?
-        next if content.include?(thread.anchor_context)
+
+      if thread.anchor_start.present? && thread.anchor_end.present? && ops_json.any? { |op| op.key?("resolved_range") || op.key?("replacements") }
+        # Position-based check: transform anchor range through the new edit
+        begin
+          new_range = Plans::TransformRange.transform_through_versions(
+            [thread.anchor_start, thread.anchor_end],
+            [new_version]
+          )
```
[the remark this answers is not in the document]
Comment (HamptonMakes) [reply app/models/comment_thread.rb]: Fixed in 5398eb1. `mark_out_of_date_for_new_version!` now queries all versions from `thread.anchor_revision` to `new_version.revision` and transforms through the ones with positional ops. When intervening versions lack positional data, falls back to text-based checking instead of silently advancing the anchor revision.

File: app/controllers/api/v1/operations_controller.rb [hunk 4]
```diff
@@ -38,13 +90,130 @@ def create
...
+        base_version = @plan.plan_versions.find_by(revision: base_revision)
+        unless base_version
+          render json: { error: "Base revision #{base_revision} not found" }, status: :conflict
+          return
+        end
+
+        intervening_versions = @plan.plan_versions
+          .where("revision > ? AND revision <= ?", base_revision, @plan.current_revision)
+          .order(revision: :asc)
+          .to_a
```
[the remark this answers is not in the document]
Comment (HamptonMakes) [reply app/controllers/api/v1/operations_controller.rb]: Fixed. `rebase_and_resolve` now checks all intervening versions for non-positional ops before attempting the rebase. If any version has operations without `resolved_range`/`replacements`, we return a 409 conflict with a clear message to re-read the plan, instead of silently skipping legacy edits and risking wrong-offset mutations.

File: app/controllers/api/v1/operations_controller.rb [hunk 5]
```diff
@@ -38,13 +90,132 @@ def create
...
+
+        operations.each do |op|
+          op = op.respond_to?(:to_unsafe_h) ? op.to_unsafe_h.transform_keys(&:to_s) : op.transform_keys(&:to_s)
+          begin
+            resolution = Plans::PositionResolver.call(content: working_base, operation: op)
+            transformed_ranges = resolution.ranges.map do |range|
+              Plans::TransformRange.transform_through_versions(range, intervening_versions)
+            end
+
+            if op["op"] == "replace_exact" && op["old_text"]
```
[the remark this answers is not in the document]
Comment (HamptonMakes) [reply app/controllers/api/v1/operations_controller.rb]: This seems legit.

**Comment by HamptonMakes** [comment]
@codex please re-review
